stores/chain.go

```diff
@@ -0,0 +1,356 @@
+		Where("fcid = ?", fileContractID(fcid)).
+		Scan(&state).
+		Error
+
+	if err == gorm.ErrRecordNotFound {
+		err = u.tx.
+			Select("state").
+			Model(&dbArchivedContract{}).
+			Where("fcid = ?", fileContractID(fcid)).
+			Scan(&state).
+			Error
+	}
+
+	if err != nil {
+		return "", err
+	}
+	return api.ContractState(state.String()), nil
+}
+
+// RemoveSiacoinElements is called with all siacoin elements that were spent in
+// the update.
+func (u *chainUpdateTx) RemoveSiacoinElements(ids []types.SiacoinOutputID) error {
+	for _, id := range ids {
+		if err := u.tx.
+			Where("output_id", hash256(id)).
+			Delete(&dbWalletOutput{}).
+			Error; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RevertIndex is called with the chain index that is being reverted. Any
+// transactions and siacoin elements that were created by the index should be
+// removed.
+func (u *chainUpdateTx) RevertIndex(index types.ChainIndex, removed, unspent []types.SiacoinElement) error {
+	// recreate unspent outputs
+	for _, e := range unspent {
+		if err := u.tx.
+			Clauses(clause.OnConflict{
+				DoNothing: true,
+				Columns:   []clause.Column{{Name: "output_id"}},
+			}).
+			Create(&dbWalletOutput{
+				OutputID:       hash256(e.ID),
+				LeafIndex:      e.StateElement.LeafIndex,
+				MerkleProof:    merkleProof{proof: e.StateElement.MerkleProof},
+				Value:          currency(e.SiacoinOutput.Value),
+				Address:        hash256(e.SiacoinOutput.Address),
+				MaturityHeight: e.MaturityHeight,
+				Height:         index.Height,
+				BlockID:        hash256(index.ID),
+			}).Error; err != nil {
+			return nil
+		}
+	}
+
+	// remove outputs created at the reverted index
+	for _, e := range removed {
+		if err := u.tx.
+			Where("output_id", hash256(e.ID)).
+			Delete(&dbWalletOutput{}).
+			Error; err != nil {
+			return err
+		}
+	}
+
+	// remove events created at the reverted index
+	return u.tx.
+		Model(&dbWalletEvent{}).
+		Where("height = ? AND block_id = ?", index.Height, hash256(index.ID)).
+		Delete(&dbWalletEvent{}).
+		Error
+}
+
+// UpdateChainIndex updates the chain index in the database.
+func (u *chainUpdateTx) UpdateChainIndex(index types.ChainIndex) error {
+	return u.tx.
+		Model(&dbConsensusInfo{}).
+		Where(&dbConsensusInfo{Model: Model{ID: consensusInfoID}}).
+		Updates(map[string]interface{}{
+			"height":   index.Height,
+			"block_id": hash256(index.ID),
+		}).
+		Error
+}
+
+// UpdateContract updates the revision height, revision number, and size the
+// contract with given fcid.
+func (u *chainUpdateTx) UpdateContract(fcid types.FileContractID, revisionHeight, revisionNumber, size uint64) error {
+	updates := map[string]interface{}{
+		"revision_height": revisionHeight,
+	}
+	if revisionNumber > 0 {
+		updates["revision_number"] = fmt.Sprint(revisionNumber)
+	}
+	if size > 0 {
+		updates["size"] = size
+	}
+
+	if err := u.tx.
+		Model(&dbContract{}).
+		Where("fcid = ?", fileContractID(fcid)).
+		Updates(updates).
```

Comment:
We should split that up into 2 updates.
Update 1: updates revision height regardless of revision number since this is only set by the subscriber anyway and that way we correctly handle reorgs that change the height as well.
Update 2: updates size and revision number if the revision number is > the one we have stored already. Size and revision number aren't affected by reorgs since they are meant to reflect the latest off-chain state.

I don't think we can do both in a single query because one of them has a condition while the other one doesn't.

Reply:
This changes in the next PR already. It was more or less what you propose aside from the fact that I was not updating the rev height in all cases.